File: contrib/sdk-bindings/rust/python-tests/test_micheline.py
# SPDX-FileCopyrightText: 2025 Functori <contact@functori.com>
#
# SPDX-License-Identifier: MIT

from tezos import (
    Prim,
    Annotation,
    Micheline,
    BigInt,
    MichelineManager,
)

def test_build_basic_script_micheline():
    """
    { parameter unit ; storage unit ; code { CDR ; NIL operation ; PAIR } }
    """

    micheline = Micheline.SEQ([
        Micheline.APP(Prim.K_PARAMETER, [Micheline.APP(Prim.T_UNIT, [], [])], []),
        Micheline.APP(Prim.K_STORAGE, [Micheline.APP(Prim.T_UNIT, [], [])], []),
        Micheline.APP(Prim.K_CODE, [
            Micheline.SEQ([
                Micheline.APP(Prim.I_CDR, [], []),
                Micheline.APP(Prim.I_NIL, [Micheline.APP(Prim.T_OPERATION, [], [])], []),
                Micheline.APP(Prim.I_PAIR, [], []),
            ])
        ], []),
    ])

    micheline_manager = MichelineManager()

    parsed_micheline = micheline_manager.parse(
        "{ parameter unit ; storage unit ; code { CDR ; NIL operation ; PAIR } }"
    )
    assert micheline_manager.equal_micheline(micheline, parsed_micheline)

def test_build_simple_data_micheline():
    """
    (Pair :foo "string" 0 0x00)
    """

    micheline = Micheline.APP(Prim.D_PAIR, [
        Micheline.STRING("string"),
        Micheline.INT(BigInt.from_int(0)),
        Micheline.BYTES(bytes.fromhex('00')),
    ], [Annotation.TYPE("foo")])

    micheline_manager = MichelineManager()

    parsed_micheline = micheline_manager.parse("""(Pair :foo "string" 0 0x00)""")
    assert micheline_manager.equal_micheline(micheline, parsed_micheline)

File: contrib/sdk-bindings/swift/Tests/TestMicheline.swift
// SPDX-FileCopyrightText: 2025 Functori <contact@functori.com>
//
// SPDX-License-Identifier: MIT

import XCTest
import Tezos

class TestMicheline: XCTestCase {

    public static var allTests = [
      ("testBuildBasicScriptMicheline", testBuildBasicScriptMicheline),
      ("testBuildSimpleDataMicheline", testBuildSimpleDataMicheline),
    ]

    /*
     { parameter unit ; storage unit ; code { CDR ; NIL operation ; PAIR } }
     */
    func testBuildBasicScriptMicheline() {
        let micheline = Micheline.seq(
          seq: [
            Micheline.app(
              prim: Prim.kParameter,
              seq: [Micheline.app(prim: Prim.tUnit, seq: [], annots: [])],
              annots: []
            ),
            Micheline.app(
              prim: Prim.kStorage,
              seq: [Micheline.app(prim: Prim.tUnit, seq: [], annots: [])],
              annots: []
            ),
            Micheline.app(
              prim: Prim.kCode,
              seq: [
                Micheline.seq(
                  seq: [
                    Micheline.app(prim: Prim.iCdr, seq: [], annots: []),
                    Micheline.app(
                      prim: Prim.iNil,
                      seq: [Micheline.app(prim: Prim.tOperation, seq: [], annots: [])],
                      annots: []
                    ),
                    Micheline.app(prim: Prim.iPair, seq: [], annots: []),
                  ])
              ],
              annots: []
            ),
          ])

        let michelineManager = MichelineManager()

        let parsedMicheline = try! michelineManager.parse(
          micheline:
            "{ parameter unit ; storage unit ; code { CDR ; NIL operation ; PAIR } }"
        )
        XCTAssertTrue(michelineManager.equalMicheline(micheline1: micheline, micheline2: parsedMicheline))
    }

    /*
     (Pair :foo "string" 0 0x00)
    */
    func testBuildSimpleDataMicheline() {
        let micheline = Micheline.app(
          prim: Prim.dPair,
          seq: [
            Micheline.string(string: "string"),
            Micheline.int(bigInt: BigInt.fromInt(num: 0)),
            Micheline.bytes(bytes: Data([0x00])),
          ],
          annots: [Annotation.type("foo")]
        )

        let michelineManager = MichelineManager()

        let parsedMicheline = try! michelineManager.parse(
          micheline: "(Pair :foo \"string\" 0 0x00)"
        )
        XCTAssertTrue(michelineManager.equalMicheline(micheline1: micheline, micheline2: parsedMicheline))
    }

}
